### PR TITLE
Multimap evolutions

### PR DESCRIPTION
**Pull request by karesti**
https://issues.redhat.com/browse/ISPN-13856

### CODE REVIEW

File: Multimap-As-A-First-Class-Data-Structure.asciidoc [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+GET /rest/v2/multimaps/
+
+GET /rest/v2/multimaps/{multimapName}?action=???
+
+DELETE /rest/v2/multimaps/{multimapName}/{multimapKey}
+----
+
+* CRUD multimap (providing a configuration to create a multimap, templates not supported)
+* Display Multimap content entries
+* Get values by key
```
Comment (tristantarrant) [inline Multimap-As-A-First-Class-Data-Structure.asciidoc:92]: We also need an equivalent of the `CacheEntry<K, V> Cache.getEntry()` call. Unfortunately this needs to have a different interface per type

```java
// sync
SyncCacheEntryCollection<K, V> MultiMap.getEntry(K key, CacheOptions options);

public interface SyncCacheEntryCollection<K, V> {
   K key();

   CloseableIterable<V> values();

   CacheEntryMetadata metadata();
}
// async
CompletionStage<AsyncCacheEntryCollection<K, V>> MultiMap.getEntry(K key, CacheOptions options);

public interface AsyncCacheEntryCollection<K, V> {
   K key();

   Flow.Publisher<V> values();

   CacheEntryMetadata metadata();
}
// mutiny
Uni<MutinyCacheEntryCollection<K, V>> MultiMap.getEntry(K key, CacheOptions options);

public interface MutinyCacheEntryCollection<K, V> {
   K key();

   Multi<V> values();

   CacheEntryMetadata metadata();
}
```

File: Multimap-As-A-First-Class-Data-Structure.asciidoc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+* In Indexing is enabled, multimaps support query
+* CRUD Key/Value. If a key exists we add the value. To remove a key/value we need to send the value as well.
+* CRUD Key/AllValues. We remove the entry completely.
+* Clearing Multimap operation
+* Enable/Disable rebalancing
+* Availability get/set
+* Replace not supported
```
Comment (tristantarrant) [inline Multimap-As-A-First-Class-Data-Structure.asciidoc:99]: I'd mention listener support as well as possible interactions with near-caching

**Review by tristantarrant — CHANGES_REQUESTED** [review]

File: Multimap-As-A-First-Class-Data-Structure.asciidoc (hunk 2 above)
Comment (karesti) [reply Multimap-As-A-First-Class-Data-Structure.asciidoc:99]: Yes, true. There was in the roadmap as well

File: Multimap-As-A-First-Class-Data-Structure.asciidoc (hunk 1 above)
Comment (wburns) [reply Multimap-As-A-First-Class-Data-Structure.asciidoc:92]: This also brings up the question do values each have their own metadata such as expiration? The API as proposed here wouldn't allow that.

Comment (karesti) [reply Multimap-As-A-First-Class-Data-Structure.asciidoc:92]: For now not. Do we want that?

Comment (karesti) [reply Multimap-As-A-First-Class-Data-Structure.asciidoc:92]: @tristantarrant @wburns how would we include that ?

**Comment by karesti** [comment]
@wburns the expiration the way we implemented multimap, can only be supported by key, not per key value. I believe being able to change it all to support expiration per key/value won't be able to land very soon. I think we might want to either support expiration of keys and all values inside or not support expiration at all for now.

**Comment by karesti** [comment]
@tristantarrant @wburns I merged this one instead of closing it

**Pull request merged**